pkg/cable/iptun/iptun.go

```diff
@@ -0,0 +1,356 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptun
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/cable"
+	"github.com/submariner-io/submariner/pkg/cable/cableutils"
+	"github.com/submariner-io/submariner/pkg/natdiscovery"
+	netlinkAPI "github.com/submariner-io/submariner/pkg/netlink"
+	"github.com/submariner-io/submariner/pkg/routeagent_driver/cni"
+	"github.com/submariner-io/submariner/pkg/types"
+	"github.com/vishvananda/netlink"
+	"k8s.io/klog"
+)
+
+const (
+	IPTunIface         = "ipip-tunnel"
+	IPTunOverhead      = 20
+	IPTunNetworkPrefix = 242
+	CableDriverName    = "iptun"
+	TableID            = 100
+)
+
+type Operation int
+
+const (
+	Add Operation = iota
+	Delete
+	Flush
+)
+
+type iptun struct {
+	localEndpoint types.SubmarinerEndpoint
+	localCluster  types.SubmarinerCluster
+	connections   []v1.Connection
+	mutex         sync.Mutex
+	ipTunIface    *ipTunIface
+	netLink       netlinkAPI.Interface
+}
+
+type ipTunIface struct {
+	activeEndpointHostname string
+	link                   *netlink.Iptun
+	iptunIP                net.IP
+}
+
+type iptunAttributes struct {
+	name     string
+	pmtuDisc uint8
+	local    net.IP
+	remote   net.IP
+	mtu      int
+}
+
+func init() {
+	cable.AddDriver(CableDriverName, NewDriver)
+}
+
+func NewDriver(localEndpoint *types.SubmarinerEndpoint, localCluster *types.SubmarinerCluster) (cable.Driver, error) {
+	// We'll panic if localEndpoint or localCluster are nil, this is intentional
+	ipip := iptun{
+		localEndpoint: *localEndpoint,
+		netLink:       netlinkAPI.New(),
+		localCluster:  *localCluster,
+	}
+
+	// Don't create an ipip-tunnel interface till you have the remote endpoint info.
+	return &ipip, nil
+}
+
+func (ipip *iptun) createIPTunInterface(remoteIP net.IP) error {
```

Comment:
It's [recommended](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names) that receiver names be short, one or two letters.

```suggestion
func (i *iptun) createIPTunInterface(remoteIP net.IP) error {
```

Reply:
Q: There are discussions about installing Submariner GW will K8S LBS, how the IPinIP will be distributed in such case? AFAIK, K8S LBS supports only L4 LB

Reply:
I will update them all